p2p/whitelist_connection_gater.go

```diff
@@ -0,0 +1,79 @@
+package p2p
+
+import (
+	"github.com/libp2p/go-libp2p/core/control"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/rs/zerolog"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+	maddr "github.com/multiformats/go-multiaddr"
+)
+
+type WhitelistConnectionGater struct {
+	whitelistedPeers map[peer.ID]bool
+	logger           zerolog.Logger
+}
+
+func NewWhitelistConnectionGater(whitelistedPeers []peer.ID, logger zerolog.Logger) *WhitelistConnectionGater {
+	gater := &WhitelistConnectionGater{
+		logger:           logger,
+		whitelistedPeers: make(map[peer.ID]bool),
+	}
+
+	for _, p := range whitelistedPeers {
+		logger.Info().Msgf("Adding peer %s to whitelist", p)
+		gater.whitelistedPeers[p] = true
+	}
+
+	return gater
+}
+
+func (wg *WhitelistConnectionGater) InterceptPeerDial(p peer.ID) (allow bool) {
+	return wg.peerAllowed("InterceptPeerDial", p, nil)
+}
+
+func (wg *WhitelistConnectionGater) InterceptAddrDial(p peer.ID, m maddr.Multiaddr) (allow bool) {
+	return wg.peerAllowed("InterceptAddrDial", p, &m)
+}
+
+func (wg *WhitelistConnectionGater) InterceptAccept(m network.ConnMultiaddrs) (allow bool) {
+	return true
+}
+
+func (wg *WhitelistConnectionGater) InterceptSecured(direction network.Direction, p peer.ID, m network.ConnMultiaddrs) (allow bool) {
+	remoteMultiAddr := m.RemoteMultiaddr()
+	return wg.peerAllowed("InterceptSecured", p, &remoteMultiAddr)
+}
+
+func (wg *WhitelistConnectionGater) InterceptUpgraded(network.Conn) (bool, control.DisconnectReason) {
+	// Allow connection upgrades
+	return true, 0
+}
+
+func (wg *WhitelistConnectionGater) peerAllowed(interceptor string, p peer.ID, remoteAddr *maddr.Multiaddr) bool {
+	allowed := wg.whitelistedPeers[p]
+
+	var event *zerolog.Event
+	if allowed {
+		event = wg.logger.Debug() // log allowed peers at Debug level
+	} else {
+		event = wg.logger.Info() // log denied peers at Info level
+	}
+
+	event = event.
+		Str("interceptor", interceptor).
+		Str("peer", p.String()).
```

Comment:
```suggestion
		Stringer("peer", p).
```

would only call `.String()` if the log level is actually enabled